spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java

```diff
@@ -422,21 +421,36 @@ protected ErrorMessageStrategy getErrorMessageStrategy() {
 	protected MessageHandler getErrorMessageHandler(final ConsumerDestination destination, final String group,
 			final ExtendedConsumerProperties<KafkaConsumerProperties> extendedConsumerProperties) {
 		if (extendedConsumerProperties.getExtension().isEnableDlq()) {
-			ProducerFactory<byte[], byte[]> producerFactory = this.transactionManager != null
+			ProducerFactory<?,?> producerFactory = this.transactionManager != null
 					? this.transactionManager.getProducerFactory()
-					: getProducerFactory(null, new ExtendedProducerProperties<>(new KafkaProducerProperties()));
-			final KafkaTemplate<byte[], byte[]> kafkaTemplate = new KafkaTemplate<>(producerFactory);
+					: getProducerFactory(null,
+						new ExtendedProducerProperties<>(extendedConsumerProperties.getExtension().getDlqProducerProperties()));
+			final KafkaTemplate<?,?> kafkaTemplate = new KafkaTemplate<>(producerFactory);
+			String dlqName = StringUtils.hasText(extendedConsumerProperties.getExtension().getDlqName())
+					? extendedConsumerProperties.getExtension().getDlqName()
+					: "error." + destination.getName() + "." + group;
+
+			@SuppressWarnings({"unchecked", "raw"})
+			DlqSender<?,?> dlqSender = new DlqSender(kafkaTemplate, dlqName);
+
 			return message -> {
 				final ConsumerRecord<?, ?> record = message.getHeaders()
 						.get(KafkaHeaders.RAW_DATA, ConsumerRecord.class);
-				final byte[] key = record.key() != null ? Utils.toArray(ByteBuffer.wrap((byte[]) record.key()))
-						: null;
-				final byte[] payload = record.value() != null
-						? Utils.toArray(ByteBuffer.wrap((byte[]) record.value()))
-						: null;
-				String dlqName = StringUtils.hasText(extendedConsumerProperties.getExtension().getDlqName())
-						? extendedConsumerProperties.getExtension().getDlqName()
-						: "error." + destination.getName() + "." + group;
+
+				if (extendedConsumerProperties.isUseNativeDecoding()) {
+					if (record != null) {
+						Map<String, String> configuration = extendedConsumerProperties.getExtension().getDlqProducerProperties().getConfiguration();
+						if (record.key() != null && !record.key().getClass().isInstance(byte[].class)) {
+							ensureDlqMessageCanBeProperlySerialized(
```

Comment:
These checks need to be done against the global producer properties if the binder is transactional - all producers (including the DLQ) have to use the producer factory registered with the transaction manager (see getting the PF above).

`configurationProperties.getTransaction().getProducer()`